onnxruntime/core/providers/dml/DmlExecutionProvider/src/MLOperatorAuthorImpl.cpp

```diff
@@ -982,6 +982,85 @@ namespace Windows::AI::MachineLearning::Adapter
         m_abiExecutionObject.CopyTo(executionInterface);
     }
 
+    uint32_t STDMETHODCALLTYPE OpKernelInfoWrapper::GetUtf8NameSizeInBytes() const noexcept
+    {
+        // Include null terminator.
+        const auto& name = m_impl->node().Name(); 
+        return name.empty() ? 0 : name.size() + 1;
+    }
+
+    HRESULT STDMETHODCALLTYPE OpKernelInfoWrapper::GetUtf8Name(uint32_t bufferSizeInBytes, char* outputName) const noexcept
+    {
+        if (bufferSizeInBytes == 0)
+        {
+            return E_INVALIDARG;
+        }
+
+        // Copy as many characters as possible, leaving room for the null terminator.
+        const auto& nodeName = m_impl->node().Name();
+        size_t charsCopied = nodeName.copy(outputName, bufferSizeInBytes - 1);
+
+        // Write the null terminator.
+        assert(charsCopied >= 0 && charsCopied < bufferSizeInBytes);
+        outputName[charsCopied] = '\0';
+
+        return S_OK;
+    }
+
+    uint32_t STDMETHODCALLTYPE OpKernelInfoWrapper::GetWideNameSizeInBytes() const noexcept
+    {
+        const auto& name = m_impl->node().Name(); 
+        if (name.empty())
+        {
+            return 0;
+        }
+        
+        int requiredSize = MultiByteToWideChar(CP_UTF8, 0, name.data(), name.size(), nullptr, 0);
+        assert(requiredSize > 0);
+
+        // Include null terminator.
+        return static_cast<uint32_t>(requiredSize) + 1;
+    }
+
+    HRESULT STDMETHODCALLTYPE OpKernelInfoWrapper::GetWideName(uint32_t bufferSizeInBytes, wchar_t* outputName) const noexcept
+    {
+        // Buffer needs to be large enough to at least hold a null terminator.
+        if (bufferSizeInBytes < sizeof(wchar_t))
+        {
+            return E_INVALIDARG;
+        }
+
+        const auto& nodeName = m_impl->node().Name();
+        if (nodeName.empty())
+        {
+            outputName[0] = L'\0';
+            return S_OK;
+        }
+
+        uint32_t bufferSizeInChars = bufferSizeInBytes / sizeof(wchar_t);
+        int charsCopiedIfSucceeded = MultiByteToWideChar(CP_UTF8, 0, nodeName.data(), nodeName.size(), outputName, bufferSizeInChars);
+
+        auto lastError = GetLastError();
```

Comment:
The return value should be checked, not last error. GetLastError should only be called *if* the return value indicates there is an error, because very many Win32 functions do not clear any remaining error previously set. They only set a new error when there was an error. So we could reach line 1044, and GetLastError could still have a residual ERROR_INSUFFICIENT_BUFFER value from a *previous* failed API call, making us think that *this* call failed.